tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/contact/OpenPaasContactsConsumer.java

```diff
@@ -0,0 +1,182 @@
+            .subscribe();
+    }
+
+    public Flux<AcknowledgableDelivery> delivery() {
+        return Flux.using(receiverProvider::createReceiver,
+            receiver -> receiver.consumeManualAck(QUEUE_NAME),
+            Receiver::close);
+    }
+
+    private Mono<EmailAddressContact> messageConsume(AcknowledgableDelivery ackDelivery, String messagePayload) {
+        return Mono.just(messagePayload)
+            .map(this::parseContactAddedRabbitMqMessage)
+            .flatMap(this::handleMessage)
+            .doOnSuccess(result -> {
+                LOGGER.warn("Consumed contact successfully '{}'", result);
+                ackDelivery.ack();
+            })
+            .onErrorResume(error -> {
+                LOGGER.error("Error when consume message '{}'", messagePayload, error);
+                ackDelivery.nack(REQUEUE_ON_NACK);
+                return Mono.empty();
+            });
+    }
+
+    private ContactAddedRabbitMqMessage parseContactAddedRabbitMqMessage(String message) {
+        try {
+            return objectMapper.readValue(message, ContactAddedRabbitMqMessage.class);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException("Failed to parse ContactAddedRabbitMqMessage", e);
+        }
+    }
+
+    private Mono<EmailAddressContact> handleMessage(ContactAddedRabbitMqMessage contactAddedMessage) {
+        LOGGER.info("Consumed jCard object message: {}", contactAddedMessage);
+        Optional<ContactFields> contactFieldsOpt = toContactFields(contactAddedMessage.vcard());
+        return openPaasRestClient.retrieveMailAddress(contactAddedMessage.userId())
+            .map(ownerMailAddress -> AccountId.fromUsername(Username.fromMailAddress(ownerMailAddress)))
+            .flatMap(ownerAccountId ->
+                Mono.justOrEmpty(contactFieldsOpt)
+                    .flatMap(contactFields -> doAddContact(ownerAccountId, contactFields)));
+    }
+
+    private Mono<EmailAddressContact> doAddContact(AccountId ownerAccountId, ContactFields contactFields) {
+        return Mono.from(contactSearchEngine.index(ownerAccountId, contactFields));
+    }
+
+    private Optional<ContactFields> toContactFields(JCardObject jCardObject) {
+        Optional<String> contactFullnameOpt = jCardObject.fnOpt();
+        Optional<MailAddress> contactMailAddressOpt = jCardObject.emailOpt()
+            .flatMap(contactEmail -> {
+                try {
+                    return Optional.of(new MailAddress(contactEmail));
+                } catch (AddressException e) {
+                    LOGGER.warn("Invalid contact email address: {}", contactEmail, e);
+                    return Optional.empty();
+                }
+            });
+
+        return combineOptionals(contactFullnameOpt, contactMailAddressOpt,
+            (contactFullname, contactMailAddress) ->
+                new ContactFields(contactMailAddress, contactFullname, contactFullname));
```

Comment:
Does it make sense to index twice the full name?

Reply:
No, it does not, I changed it to index the fullname as the firstname and leave the surname empty. We can't just split by space the full name and put one part as a first name and the other as a surname as some fullnames could contain a prefix like 'Dr.' which would make the fullname containing more than one space.